remove cmsis module; add acle module

japaric opened the pull request:
ACLE (ARM C Language Extensions) is more general (supports ARMv4 to ARMv8) than
CMSIS (ARMv7-M and ARMv7-R)

Previous discussion: https://github.com/rust-embedded/wg/pull/184

TODO in follow up PRs:

- Move `src/arm/dsp.rs` into `src/acle/{dsp,simd32}.rs`. Also the names used in `arm::dsp` don't match ACLE -- they should start with `__`. cc @paoloteti 

- Add 64-bit system registers. cc @andre-richter

Pre-requisite: whitelist the aclass, v5te, v6k and v6t2 target features in rustc.

closes #437 

r? @gnzlbg

japaric commented:
> Pre-requisite: whitelist the aclass, v5te, v6k and v6t2 target features in rustc

See rust-lang/rust#53926

@gnzlbg 

> I think the way the registers are extended with traits and used is a bit opinionated. I would like to see how the C code compares to the Rust code for some very simple tasks, like just reading the value of a register.

ACLE uses strings. For example usage, this is [how CMSIS is implemented using ACLE](https://github.com/ARM-software/CMSIS_5/blob/f95e113f017437d1357417500e48f51009cf07ee/CMSIS/Core/Include/cmsis_iccarm.h#L270). We could use strings but they are error prone (e.g. what should `__rsr("NOT_A_REGISTER")` and `__rsr("ASPR")` (that's a typo; it's "APSR") do? panic?) and IDK if `rustc_args_required_const` supports string literals?

japaric commented:
@gnzlbg CI is now green.

gnzlbg commented:
@alexcrichton this direction is pretty much the consensus of the embedded-wg and this looks good to me. Do you want to / have time to review this?

parched commented:
@gnzlbg 
> Is there a "better" implementation of the YIELD intrinsic (e.g. using llvm intrinsics) than using inline assembly?

You can check what IR clang lowers these to, in this case it's `llvm.arm.hint(1)`. `spin_loop_hint` does just use inline asm though. My question is, does this need to be added here if it's already available in rust? In general I think it's wise to only add things here that have a reasonable use case to reduce the maintenance and chance for error.

> We should keep all intrinsics `unsafe` for the time being, and if someone wants to relax that for some intrinsics, they can do so in a backwards compatible way in a follow up RFC after the main work is done.

Ok, that seems reasonable to me.

> @parched sadly core/std are extremely explicit in that architecture-specific functionality is always behind a module containing the architecture name (arm/aarch64). ...

Ok, that's fair enough.

gnzlbg commented:
@parched 

> You can check what IR clang lowers these to, in this case it's llvm.arm.hint(1). spin_loop_hint does just use inline asm though. My question is, does this need to be added here if it's already available in rust? In general I think it's wise to only add things here that have a reasonable use case to reduce the maintenance and chance for error.

That makes sense. I personally think this intrinsic belongs here, and that spin lock hint should just be upgraded to call it once we land this. I don't know if this intrinsic is only useful for spin lock hint, or if it potentially can have other uses. If its the later, then putting it here makes it more "discoverable".

andre-richter commented:
The first time I got a chance to look at this.
I think I like how the shareability domains are structs that implement barrier instructions as traits. Did you come up with that for especially this PR or is this scheme used elsewhere already?

May I ask in which place it will be documented how to properly use this ACLE implementation? For example that using dsb() can/has to be combined with the provided shareability domain structs might not be clear until you browse through all the code.

Or will we spare this effort here because ideally it's us who wrap this right away into the cortex-m/r/a crates and elaborately document there?

japaric commented:
Meta-comment: The embedded WG has removed "stable core::arch::arm" from the 2018 milestone since we have a good enough stable `global_asm!` hack that doesn't have a build dependency on (arm-none-eabi-)GCC. We'd still love to have these intrinsics because they can inline into user code but it's no longer a priority for this year.

---

@parched

> How about using an enum for the barrier parameter, rather than the specialization of some generic parameter?

What's the motivation? The current approach is extensible and doesn't depend on unimplemented features.

Using an enum means that, to account for future additions (e.g. ARMv8.5 gets a new shareability domain), a hidden variant will have to be added to it. Which means that intrinsics that use the enum will have to include a match arm that panics on use of the hidden variant. Which means that unoptimized calls to the intrinsic will include the panicking branch, which could pull in the formatting machinery bloating the binary by a few KB.

If generics are so undesirable I would prefer to directly expose this:

``` rust
#[rustc_args_required_const(0)]
pub fn __dmb(val: i32) {
    extern "C" {
        #[link_name = "llvm.arm.dmb"]
        fn __dmb(val: i32);
    }

    __dmb(val)
}
```

It errors at compile time ("LLVM ERROR: Cannot select: intrinsic %llvm.arm.dmb") when you use an invalid value (e.g. greater than 16).

> Unfortunately rsr/wsr must take a &str parameter to allow for arbitrary "S_x_x_x_x" register names in future. (..).

Could you give examples of these *arbitrary* registers? I'm aware of registers of the form `cp<coprocessor>:<opc1>:c<CRn>:c<CRm>:<opc2>` where `<coprocessor>` and the other placeholders can take integer values from 0 to 7 or 0 to 15. Are you referring to those or to something else? (I wouldn't use the word "arbitrary" for those; there's a known number of them defined by ARM; "arbitrary" to me sounds like some ARM licensee can come up with any random register name and then only their toolchain would support them)

> In future when compile time arguments are supported the match can be removed and all strings allowed.

Could you elaborate how compile time arguments let us specify arbitrary register names w/o a match? To me the solution would require string interpolation via a macro (see below).

``` rust
macro_rules! __rsr {
    ($register:expr) => {
        {
            let r: u32;
            asm!(concat!("mrs $0,", $register) : "=r"(r) : : : "volatile");
            r
        }
    }
}
```

> Implementation of these intrinsics should probably use llvm.{arm,aarch64}.{hint,dmb,dsb.isb} and llvm.{read,write}_register

SGTM but the LLVM docs say about llvm.{read,write}_register:

> Warning: So far it only works with the stack pointer on selected architectures (ARM, AArch64, PowerPC and x86_64). Significant amount of work is needed to support other registers and even more so, allocatable registers.

Which mean I can't use those two in this PR.

---

@gnzlbg

> what do you think? It might be worth it to just expose here a "stringy"-based API like ACLE does on C, and just expose an idiomatic API on top of it in crates.io.

I would prefer to avoid matches with panicking branches in this API because then anything built on top of this API will inherit them. Sure, the panicking branches will be optimized away in release but unoptimized code will keep them and will be bigger than it needs to be (because of the formatting machinery).

If we could get the miri in the compiler to pre-process functions that have the `rustc_args_required_const` attribute and eliminate all the unreachable branches then I'd be totally OK with matches and stringy APIs.

---

Given that we are not time pressured I would personally prefer to optimize this API to be as slim as inline `asm!` regardless of the applied optimization level.

paoloteti commented:
@japaric 

> Warning: So far it only works with the stack pointer on selected architectures (ARM, AArch64, PowerPC and x86_64). Significant amount of work is needed to support other registers and even more so, allocatable registers.

Probably this remark is outdated. I'm sure `llvm.arm.write_register` support other registers (probably all), but it is not a function you can link using `link_name=`. You need to write few lines of LLVM IR (at least in LLVM v6.0.x). `clang` use the LLVM API to generate this IR and its metadata.

parched commented:
@japaric 

> What's the motivation? The current approach is extensible and doesn't depend on unimplemented features.

It just seems a bit more intuitive and nicer use to me, like  `atomic::fence`'s memory order parameter, and it closer matches the semantics of the instruction. What advantage does the generic approach have over explicit `dmb_ish`, `dmb_osh` etc.? 

> Which means that unoptimized calls to the intrinsic will include the panicking branch, which could pull in the formatting machinery bloating the binary by a few KB.

Yes, that's unfortunate, (`sync::atomic` suffers the same problem), I guess I was hoping in the future with const arguments that could be removed or if the user really cared they would be using an optimised build.

> If generics are so undesirable I would prefer to directly expose this:

That could be fine, especially if the goal here is to be completely low level without any user friendliness. Although, I would prefer user friendliness if possible.

> Could you give examples of these arbitrary registers?

I mean the likes of implementation defined registers which rustc/LLVM will never be able to know all the names of, e.g., [L2CTLR_EL1 on the Cortex-A73](https://developer.arm.com/docs/100048/latest/system-control/aarch64-register-descriptions/l2-control-register) which is `S3_1_C11_C0_2`.

> Could you elaborate how compile time arguments let us specify arbitrary register names w/o a match.

I was thinking by just delegating to `llvm.read_register` but @paoloteti mentions that might not be possible? Did you try it with `.i32`/`.i64` suffix @paoloteti?

paoloteti commented:
@parched 

> I was thinking by just delegating to `llvm.read_register` but @paoloteti mentions that might not be possible? Did you try it with `.i32`/`.i64` suffix @paoloteti?

`llvm.read_register.i32 ` needs a "Type Metadata" as parameter and I know only two ways to generate metadata: using the LLVM Metadata class (as `clang` does) or with few lines of IR:

```
define i32 @read_apsr() nounwind {
entry:
  %0 = call i32 @llvm.read_register.i32(metadata !0)
  ret i32 %0
}
declare i32 @llvm.read_register.i32(metadata) nounwind
!0 = !{!"apsr"}
```
Previous IR code generate `mrs r0, apsr`.

By the way, due to the signature issue, if I try with #[link_name= ..] I get
 
```
Intrinsic has incorrect argument type!                                          
LLVM ERROR: Broken function found, compilation aborted! 

```

parched commented:
Ah I see, I didn't notice the parameter was metadata.  Looks like `read/write_register` would need special handling by the compiler then.

gnzlbg commented:
The only thing I am still unsure of is whether to use the `str` API that ACLE uses or the one that this PR uses. Initially, the one that @japaric suggested sounded good, but then @parched mentioned 

> Unfortunately rsr/wsr must take a &str parameter to allow for arbitrary "S_x_x_x_x" register names in future. 

AFAICT we would need to handle those names in the intrinsic internals anyways, and if the intrinsic implementation does not support that, then newer register names won't work. I would prefer if we keep the approach that @japaric proposes here, since that would produce a compilation error in this case (e.g. because the register type is not available). This means we would need to add new types to support newer registers, but the implementation has to be upgraded anyways for those to work.

paoloteti commented:
@gnzlbg 

> I would prefer if we keep the approach that @japaric proposes here, since that would produce a compilation error in this case (e.g. because the register type is not available).

It is quite the same. The **constant** string used by `clang` in `rsr/wsr` generate a metadata entry.
For example `!0 = !{!"apsr"}` . If the string `apsr` is unknown to LLVM you still have an error.
The term 'arbitrary name' is not correct at all: string/name shall be known to LLVM in any case. 
`clang` just check that the string is constant (inlined) and manage `rsr/wsr` as a real language extention/builtin instead here We are speaking about ACLE as a library.
I use ARM Compiler day-by-day and I'm more incline to @parched suggestion that sound more familiar to me, but it is just a personal choice.

parched commented:
> AFAICT we would need to handle those names in the intrinsic internals anyways, and if the intrinsic implementation does not support that, then newer register names won't work. I would prefer if we keep the approach that @japaric proposes here, since that would produce a compilation error in this case (e.g. because the register type is not available). This means we would need to add new types to support newer registers, but the implementation has to be upgraded anyways for those to work.

The AArch64 system registers take the form `S[2-3]_[0-7]_[0-15]_[0-15]_[0-7]` which mean there's 32768 of them if I've done my maths right. Rust won't need to support all of these but it will need to support the few hundred architecturally defined ones as well as the 2048 reserved for implementation definitions. 

Ideally the intrinsic would validate they take the form of an architecturally defined name or the pattern above before letting LLVM error but that will likely depend on some detailed const argument support in rust. In C these instrinsics are normally compiler builtins so are not limited to what the language itself can do, could we do that here?

gnzlbg commented:
> In C these instrinsics are normally compiler builtins so are not limited to what the language itself can do, could we do that here?

We can't do this in `stdsimd`, but we can add a rustc intrinsic that does this, and expose it in stable Rust from `core::arch`. That way we could require the argument to be a compile-time constant, validate it at compile-time without relying on LLVM to optimize match statements, and even produce type checking errors instead of monomorphization time errors which is what a stringly-typed solution in stdsimd would produce (stable Rust does not have any monomorphization time errors right now, and we shouldn't introduce any here in anything that we plan to stabilize).

paoloteti commented:
@gnzlbg Just to summarize: basically you are saying to extend `arm.json` in `platform-intrinsics` to generate IR for `read/write_register` ?

gnzlbg commented:
So what's blocking this ? 

IIUC the only open issue is that we are not sure of whether this API or a string-based API would be better right ? It appears that this API is good for most common use cases, while the string-based API is necessary for full generality. 

Is that a correct summary of the current situation? @japaric @parched ?

gnzlbg commented:
One example that arised in https://github.com/rust-lang-nursery/stdsimd/pull/611 is that we'd like to do:

```rust
unsafe { asm!("mrs $0, S3_0_C0_C6_0" : "=r"(aa64isar0)); }
```

(or with `ID_AA64ISAR0_EL1`) but the API proposed here does not allow that.

Amanieu commented on coresimd/acle/mod.rs:
```diff
@@ -0,0 +1,140 @@
+    // v6-M
+    target_feature = "mclass"
+))]
+mod barrier;
```
According to the ACLE, these intrinsics are supported on all targets, but generate CP15 barrier instructions when the ISB/DMB/DSB instructions are not supported.

ISB: `mcr p15, 0, r0, c7, c5, 4`
DMB: `mcr p15, 0, r0, c7, c10, 5`
DSB: `mcr p15, 0, r0, c7, c10, 4`
(these are always SY scope)

The CP15 instructions are supported since the earliest ARM versions, so you don't need a feature check here.

Amanieu replied:
Also, you should change the ISB/DSB/DMB implementations to use the `llvm.arm.{isb,dsb,dmb}` intrinsics when the ISB/DSB/DMB instructions are available, and only fall back to inline asm for the CP15 intructions.

Amanieu commented on coresimd/acle/mod.rs:
```diff
@@ -0,0 +1,140 @@
+    target_feature = "rclass",
+    // v7E-M
+    all(target_feature = "mclass", target_feature = "dsp")
+))]
```
You need to guard against `target_arch = aarch64` here.

japaric replied:
I have done so but `aarch64-unknown-linux-gnu` doesn't have any of these target features: v5te, v6, mclass.

Amanieu commented on coresimd/acle/mod.rs:
```diff
@@ -0,0 +1,140 @@
+    all(target_feature = "v5te", not(target_feature = "mclass"), not(target_feature = "aclass")),
+    // v7E-M
+    all(target_feature = "mclass", target_feature = "dsp"),
+))]
```
You need to guard against `target_arch = aarch64` here.

Amanieu commented on coresimd/arm/mod.rs:
```diff
@@ -20,6 +15,7 @@ mod v7;
 #[cfg(any(target_arch = "aarch64", target_feature = "v7"))]
 pub use self::v7::*;
 
+// TODO move into the `acle::{dsp,simd32}` modules
```
Can you perform this move as part of this PR?

japaric replied:
Will do later.

Amanieu commented on coresimd/acle/simd32.rs:
```diff
@@ -0,0 +1,60 @@
+//! - __smuad
+//! - __smuadx
+//! - __smusd
+//! - __smusdx
```
You need to add a separate module which just contains `__usat` and `__ssat`. These two instructions are supported on ARMv6+ for A profile and ARMv7-M (as part of the base v7, not v7EM).

You don't need to fill in the implementation immediately, a placeholder like above would be enough.

gnzlbg commented:
This needs rebasing :)

japaric commented on coresimd/acle/hints.rs:
```diff
@@ -0,0 +1,115 @@
+/// instruction. In a multiprocessor system, it is not required to affect the
+/// other processors.
+// LLVM says "instruction requires: armv8"
+#[cfg(target_arch = "aarch64")]
```
What's the cfg for 32-bit ARMv8? There's no target_arch = "aarch32" and AFAICT  we don't have any 32-bit v8-a target listed in `rustc --print target-list`

llvm / rustc seem to be OK with emitting a SEVL instruction for thumbv7m-none-eabi and thumbv8m.main-none-eabi though

Amanieu replied:
You can use the "v8" target feature, it works the same way as the "v7" feature.

LLVM treats SEVL as an alias for the HINT instruction (`HINT #5`) and doesn't actually check if the given hint code is supported by the current target. However this is mostly harmless since unrecognized hints are treated as NOPs by the architecture.

japaric replied:
Done. We'll need to whitelist the v8 on arch=aarch64. I don't know if that feature is ever enabled on arch=arm -- it's not set for `thumbv8m.*-none-eabi*`; I would have to check if it's set for ARMv8-R.

japaric replied:
> it (v8) is not set for `thumbv8m.*-none-eabi*`

I should add that *is* correct; there's no SEVL on ARMv8-M.

> I would have to check if it's set for ARMv8-R

I checked with an hypothetical armv8r-none-eabi target and yes v8 is set. SEVL is available on ARMv8-R so this makes sense.

``` console
$ rustc +master --print cfg --target armv8r-none-eabi.json | grep feature
target_feature="dsp"
target_feature="neon"
target_feature="rclass"
target_feature="v5te"
target_feature="v6"
target_feature="v6k"
target_feature="v6t2"
target_feature="v7"
target_feature="v8"
target_feature="vfp2"
target_feature="vfp3"
target_feature="vfp4"
```

japaric replied:
> We'll need to whitelist the v8 on arch=aarch64

Err, actually `+v8` is not a valid aarch64 feature. It doesn't show up in `llc -march=aarch64 -mattr=help`. There's `+v8.1a` and others but they are not set by default for the aarch64-unknown-linux-gnu target.

japaric replied:
I have updated the cfg to include arch=aarch64.

japaric commented on coresimd/acle/mod.rs:
```diff
@@ -0,0 +1,140 @@
+//! architectures; the list includes the `target_feature`s they possess.
+//!
+//! - `armv4t-unknown-linux-gnueabi` - **ARMv4** - `+v4t`
+//! - `armv5te-unknown-linux-gnueabi` - **ARMv5TE** - `+v4t +v5te`
```
`rustc --print cfg` says otherwise

``` console
$ rustc --print cfg --target armv5te-unknown-linux-gnueabi | grep feature
target_feature="v5te"

$ rustc --print cfg --target thumbv7em-none-eabi | grep feature
target_feature="dsp"
target_feature="mclass"
target_feature="v5te"
target_feature="v6"
target_feature="v6k"
target_feature="v6t2"
target_feature="v7"
```

All the info written here corresponds to what `rustc --print cfg` outputs

Amanieu replied:
Hmm. LLVM's definition of "dsp" doesn't seem to match ACLE's definition. From my point of view this looks like an LLVM bug. Let's just work around it for now by using `(v5te && !mclass) || dsp`.

> __ARM_FEATURE_DSP is defined to 1 if the DSP (v5E) instructions are supported and the intrinsics defined in 9.4 are available. These instructions include QADD, SMULBB etc. This feature also implies support for the Q flag.

japaric commented on coresimd/acle/mod.rs:
```diff
@@ -0,0 +1,140 @@
+
+// Supported arches: 5TE, 7E-M. See Section 10.1 of ACLE (e.g. QADD)
+// But we also exclude the A profile because DSP is deprecated on that profile as of ACLE 2.0 (see
+// section 5.4.7)
```
The "dsp" target feature is not set for `arm-unknown-linux-gnueabi` though.

``` console
$ rustc --print cfg --target arm-unknown-linux-gnueabi | grep feature
target_feature="v5te"
```

Even though `asm!("QADD (..)" (..));` and others works fine.

I have tweaked the cfg to include `arm-unknown-linux-gnueabi` and `armv7-unknown-linux-gnueabihf`

japaric commented:
I have rebased this PR and addressed most of the review comments.

japaric commented:
And since we are exposing dsp / simd32 on the a-profile which is considered deprecated according to ACLE. Can we also expose LDREX / STREX? Latest ACLE has deprecated them in favor of C11/C++11 atomics, but I have them useful for optimizing lock-free code.

japaric commented:
Also we need to whitelist the 'v6' target feature in rustc since we are using it in this PR.

Amanieu commented:
I am happy with exposing ldrex/strex/clrex, since they were previously in the ACLE.

gnzlbg commented on coresimd/acle/hints.rs:
```diff
@@ -0,0 +1,115 @@
+        12 => call!(12),
+        13 => call!(13),
+        14 => call!(14),
+        _ => call!(15),
```
@Amanieu 


The Intel spec (and the assembly instructions) often state something like: "the higher bits are ignored". 

So what does the ARM spec say? What do the assembly instructions do?

If the behavior is undefined, we can emit a diagnostic (probably better to open an issue and tackle in a follow up PR). But is the higher bits or similar must be ignored, then that is what we have to do.

Amanieu replied:
These are immediate operands to the instructions. Out-of-range values result in an assembler error. Or in the case of intrinsics, a compiler error.

gnzlbg replied:
Can we expand this to `compile_error!`  in the out-of-bounds case or does that always error? 

We could also extend `#[rustc_args_required_const]` with a range to error if the const if out-of-range, e.g., `#[rustc_args_required_const(0 = "[0, 16]")]` or similar. 

What we probably want in the future is something like:

```rust
fn foo(const imm4: i32) where imm4 >= 0 && imm4 < 16 { ... }
```

or something like that. I thought that we could use an `Imm4` type similar to `NonZero` (using the `rustc_` niche start/end attributes to specify the value range, but I am not sure if assigning an invalid value causes a diagnostic:

```rust
#![feature(rustc_attrs)]
#[rustc_layout_scalar_valid_range_start(0)]
#[rustc_layout_scalar_valid_range_end(15)]
pub struct Imm4(pub u8);

fn main() {
    let _y = unsafe { Imm4(3_u8) };
    let _z = unsafe { Imm4(22_u8) };
}
```

cc @eddyb that code produces no warning, and it does not make miri fail: https://github.com/solson/miri/issues/633

gnzlbg commented:
> And since we are exposing dsp / simd32 on the a-profile which is considered deprecated according to ACLE. Can we also expose LDREX / STREX? Latest ACLE has deprecated them in favor of C11/C++11 atomics, but I have them useful for optimizing lock-free code.

We can expose them. 

> Also we need to whitelist the 'v6' target feature in rustc since we are using it in this PR.

There is a `v6k` feature white-listed, is that different from the `v6` one? I've opened a PR upstream: https://github.com/rust-lang/rust/pull/58440

gnzlbg commented:
@Amanieu what are your thoughts about the "`&str`"-API provided by ACLE vs the API proposed in this PR (e.g. https://github.com/rust-lang-nursery/stdsimd/pull/557#issuecomment-422294990). 

This is the only issue that is critical to resolve now, since it affects the API design substantially.

Amanieu commented:
I'm somewhat split on the issue. It is possible for us to support this, but it would require a bit of magic in the compiler itself.

Or we could just provide a few built-in registers and point anyone requiring more to use inline asm instead.

japaric commented:
I have addressed all the latest review comments and added ldrex / strex for arm.

However, I have not exposed ldrex / strex for aarch64 because (a) `llvm.aarch64.ldrex` doesn't lower to the right thing and (b) `asm!("LDREX")` produces an assembler / LLVM error.

I have yet to add ldrexd / strexd (64-bit version) for 32-bit arm. It's OK to return `u64` in that case, or should I instead return `(u32, u32)` which is what the llvm intrinsic returns (`{ u32, u32 }`)?

japaric commented:
I have moved the arm/dsp module into the acle/{dsp,simd32} modules

Amanieu commented:
For ldrexd / strexd just return a `u64` directly. This is what the old ACLE did.

Otherwise this looks good, but you need to fix a few CI failures.

japaric commented:
One more thing that needs to be decided for this PR is whether the API should keep the double underscore prefix that ACLE uses. This PR uses the ACLE naming scheme (e.g. `__wfi`), but the DSP / simd32 chunks of ACLE that were already in tree don't (I moved those around but didn't change the names).

We should pick one or the other. Personally, I have no preference.

Some quick statistics / facts:

- the only thing in src/arm that uses double underscores is `__breakpoint`
- there are 6 functions in src/arm that start with a *single* underscore
- there are 8 functions in src/aarch64 that start with `__crc32`
- there are 5 functions in src/aarch64 that start with a *single* underscore
- everything else in src/{arm,aarch64} doesn't use leading underscores
- the neon intrinsics specified in the "ARM Neon Intrinsics Reference" don't use leading underscores.

gnzlbg commented:
> 

I think that the easiest thing is to just do exactly what C does here. Even if that's inconsistent, it might help users to find online documentation and resources relatively easy for these APIs, even if such resources do not exist for Rust.  

Having said this, I have no strong opinion about naming these APIs, and we don't need a consistent naming for this PR to be merged.  

To stabilize this we will need an RFC, and on each aspect in which that RFC diverges from "the spec" it will need to make a case about it.

japaric commented:
I'm going to punt on ldrexd / strexd. The 'llvm.arm.strexd' and 'llvm.arm.ldrexd' intrinsics do nothing and I have no idea who to write an asm! block for `LDREXD Rx, Ry, Rz` such that `x` is even and `y = x + 1` (similar thing with STREXD).

japaric commented:
I have added underscores to the existing dsp and simd32 API to match the ACLE spec (cc @paoloteti).

If there are no more concerns / comments can we land this nightly?

Amanieu commented:
@japaric There are CI failures on the arm targets, you should fix them before we can merge this.

gnzlbg commented:
> If there are no more concerns / comments can we land this nightly?

I'll merge this as soon as Travis-CI is green (please ping me if I do not immediately do so, I do not get automatically pinged when that happens). 

We can iterate on the API issues (string-API supporting all registers vs no-string-APIs supporting only some) in tree.

mati865 commented on crates/core_arch/src/acle/simd32.rs:
```diff
@@ -386,7 +419,7 @@ pub unsafe fn usad8(a: int8x4_t, b: int8x4_t) -> u32 {
 #[cfg_attr(test, assert_instr(usad8))]
-pub unsafe fn usad8a(a: int8x4_t, b: int8x4_t, c: u32) -> u32 {
+pub unsafe fn __usada8(a: int8x4_t, b: int8x4_t, c: u32) -> u32 {
     usad8(a, b) + c
```
```suggestion
    __usad8(a, b) + c
```

mati865 commented:
Yet another error:
```
error[E0433]: failed to resolve: use of undeclared type or module `dsp`
   --> crates/core_arch/src/acle/simd32.rs:436:24
    |
436 |             assert_eq!(dsp::__qadd(-10, 60), 50);
    |                        ^^^ use of undeclared type or module `dsp`
```

japaric commented:
I have fixed most tests; now I'm seeing 8 errors like this one with `assert_instr`:

``` text
---- core_arch::acle::simd32::assert___shsub8_shsub8 stdout ----
disassembly for shsub8_shim_shsub8:
         0: push {r4, r5, r6, r7, r8, lr}
         1: ldrb r5, [sp, #36] ; 0x24
         2: uxtb r2, r2
         3: ldrb lr, [sp, #28]
         4: uxtb r0, r0
         5: ldrb r6, [sp, #32]
         6: orr r2, r2, r3, lsl #8
         7: ldrb r8, [sp, #24]
         8: orr r0, r0, r1, lsl #8
         9: orr r1, r6, r5, lsl #8
        10: ldr ip, [pc, #56] ; 7eac <__shsub8_shim_shsub8+0x68>
        11: ldr r4, [pc, #56] ; 7eb0 <__shsub8_shim_shsub8+0x6c>
        12: pkhbt r0, r0, r2, lsl #16
        13: orr r2, r8, lr, lsl #8
        14: mov r7, #42 ; 0x2a
        15: pkhbt r1, r2, r1, lsl #16
        16: add ip, pc, ip
        17: shsub8 r0, r0, r1
        18: ldr r4, [pc, r4]
        19: uxth r1, r0
        20: lsr r2, r0, #16
        21: lsr r3, r0, #24
        22: lsr r1, r1, #8
        23: str ip, [r4]
        24: str r7, [r4, #4]
        25: pop {r4, r5, r6, r7, r8, pc}
        26: .word 0x0007100b
        27: .word 0x0009612c
thread 'core_arch::acle::simd32::assert___shsub8_shsub8' panicked at 'instruction found, but the disassembly contains too many instructions: #instructions = 28 >= 22 (limit)', crates/stdsimd-test/src/lib.rs:188:9
```

The instruction we want is there but there's a bunch of other stuff around. Should I increase the instruction limit to make these tests pass?

gnzlbg commented:
You can increase the limit for particular instructions here: https://github.com/rust-lang-nursery/stdsimd/blob/master/crates/stdsimd-test/src/lib.rs#L135

japaric commented:
@gnzlbg Travis is green

gnzlbg commented:
Thanks a lot everybody! As mentioned, we can continue to iterate on this in tree as we gain experience with it.

---

I've released `core_arch` version `0.1.4` on crates.io with this changes so that you can start using this from crates.io right away - you can re-compile that easily with whatever target-features you want without having to use `Xargo` to re-compile libcore. 

I'll send a PR upstream to update stdsimd to a version including this ASAP so that one can use it on nightly, but that always takes a while. 

---

@japaric and @rust-lang-nursery/embedded-wg : after drafting an RFC for this and achieving consensus within the embedded WG on it, it might be wise to achieve consensus on the draft with some of the interested parties that have shown up in these issues and PRs before submitting a final RFC.
